R/epiparam_utils.R

```diff
@@ -438,14 +438,113 @@ as_epiparam <- function(x) {
   eparam
 }
 
-#' Subsetting method for `epiparam` class that fails when epiparam object is
-#' invalidated
+#' Subsetting method for `epiparam`
+#'
+#' @description If the subsetting invalidates the `epiparam` object (defined by
+#' its invariants, and encoded in [`validate_epiparam()`]) the subsetting will
+#' return a data frame with a message to console stating the class of the object
+#' has been converted to `data.frame` with the other attributes of the class
+#' preserved.
 #'
 #' @param epiparam An `epiparam` object
 #' @inheritParams base::subset
 #'
-#' @return An `epiparam` object
+#' @return An `epiparam` object or a `data.frame`
 #' @export
 `[.epiparam` <- function(epiparam, ...) {
-  validate_epiparam(NextMethod())
+  out <- NextMethod()
+  epiparam_reconstruct(out, epiparam)
+}
+
+#' Set names on `epiparam` class
+#'
+#' @description If the modifying the names invalidates the `epiparam` object
+#' (defined by its invariants, and encoded in [`validate_epiparam()`]) the
+#' subsetting will return a data frame with a message to console stating the
+#' class of the object has been converted to `data.frame` with the other
+#' attributes of the class preserved.
+#'
+#' @inheritParams base::names
+#'
+#' @return An `epiparam` object or a `data.frame`
+#' @export
+`names<-.epiparam` <- function(x, value) {
+  out <- NextMethod()
+  epiparam_reconstruct(out, x)
+}
+
+#' Set accessor for `epiparam` class
+#'
+#' @param x An `epiparam` object
+#' @inheritParams base::Extract
+#'
+#' @return An `epiparam` object or a `data.frame`
+#' @export
+`$<-.epiparam` <- function(x, name, value) {
+  out <- NextMethod()
+  epiparam_reconstruct(out, x)
+}
+
+#' Decides whether `epiparam` object can be reconstructed from input
+#'
+#' @description Uses [`epiparam_can_reconstruct()`] to determine whether the
+#' data input can be reconstructed in a valid `epiparam` object. If it can not,
+#' it is returned as a `data.frame`.
+#'
+#' @param x A `data.frame` or subclass of `data.frame` (e.g. `tibble` or
+#' `epiparam`)
+#' @param to The reference object, in this case an `epiparam` object
+#'
+#' @return An `epiparam` object (if the input is valid) or a `data.frame`
+#' @keywords internal
+epiparam_reconstruct <- function(x, to) {
+  if (epiparam_can_reconstruct(x)) {
+    df_reconstruct(x, to)
+  } else {
+    x <- as.data.frame(x)
+    message("Removing crucial column in `<epiparam>` returning `<data.frame>`")
```

Comment:
This could potentially be bumped up to a warning.

Reply:
In tidyverse and tidymodels packages (e.g. [^1]) this conversion is actually silent. I don't think it warrants a warning because it may be the users intention to subset out a certain column which would trigger this class downgrading. I'll leave it for now and if consensus from users builds that this should be a warning then I'll change in the future.

[^1]: https://github.com/tidymodels/dials/blob/8de943905ef0c5bcedcc7154a86c239623957d46/R/compat-vctrs-helpers-parameters.R

Reply:
That's fair, and I leave it to you - I don't have a good sense of whether users would be confused by class downgrading or take it in their stride, such as when they pipe a downgraded `epiparam` to an `epiparam` method and can't identify what causes the resulting error - as you say probably good to leave it for now and if this comes up frequently as an issue, to take another look.